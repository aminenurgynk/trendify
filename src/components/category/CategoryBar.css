
.ButtonGroup-root {
    border-radius: 20px; 
    overflow: hidden;
    box-shadow: 0 2px 12px rgba(0, 0, 0, 0.1); 
  }
  
  .Button-root {
    transition: background-color 0.3s ease;
  }
  