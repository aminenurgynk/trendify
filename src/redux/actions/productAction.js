import { ActionTypes } from "../constants/actionTypes";
import * as apiService from '../../services/apiService';


export const fetchProducts = (category = '') => async (dispatch) => {
  try {
    const endpoint = category ? `/products/category/${category}` : '/products';
    const products = await apiService.fetchProducts(endpoint);
    dispatch(setProducts(products));
  } catch (error) {
    console.error("Error fetching products:", error);
  }
};
export const setProducts = (products) => {
  return {
    type: ActionTypes.SET_PRODUCTS,
    payload: products,
  };
};

export const selectedProduct = (product) => {
  return {
    type: ActionTypes.SELECTED_PRODUCT,
    payload: product,
  };
};

export const removeSelectedProduct = () => {
  return {
    type: ActionTypes.REMOVE_SELECTED_PRODUCT,
  };
};

export const fetchCategories = () => async (dispatch) => {
  try {
    const categories = await apiService.fetchCategories();
    dispatch(setCategories(categories));
  } catch (error) {
    console.error("Error fetching categories:", error);
  }
};

export const setCategories = (categories) => {
  return {
    type: ActionTypes.SET_CATEGORIES,
    payload: categories,
  };
};

