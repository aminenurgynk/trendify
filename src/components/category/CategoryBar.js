// src/components/category/CategoryBar.js
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchCategories, fetchProducts } from '../../redux/actions/productAction';
import { ButtonGroup, Button, Container } from '@mui/material';
import './CategoryBar.css'; 

const CategoryBar = () => {
  const dispatch = useDispatch();
  const categories = useSelector((state) => state.allProducts.categories);

  useEffect(() => {
    dispatch(fetchCategories());
  }, [dispatch]);

  return (
    <Container sx={{ my: 2 }}>
      <ButtonGroup variant="text" aria-label="text button group" sx={{ display: 'flex', justifyContent: 'center', flexWrap: 'wrap' }}>
        <Button 
          sx={{
            color: 'text.primary',
            backgroundColor: 'background.paper',
            '&:hover': {
              backgroundColor: 'background.paper',
              color: 'text.primary'
            },
            margin: '8px', 
          }}
          onClick={() => dispatch(fetchProducts())}
        >
          All
        </Button>
        {categories.map((category) => (
          <Button
            key={category}
            sx={{
              color: 'text.primary',
              backgroundColor: 'background.paper',
              '&:hover': {
                backgroundColor: 'background.paper',
                color: 'text.primary'
              },
              margin: '8px', 
            }}
            onClick={() => dispatch(fetchProducts(category))}
          >
            {category}
          </Button>
        ))}
      </ButtonGroup>
    </Container>
  );
};

export default CategoryBar;
