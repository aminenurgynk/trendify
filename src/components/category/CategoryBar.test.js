import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CategoryBar from './CategoryBar';
import { Provider } from 'react-redux';
import store from '../../redux/store'; 
import { fetchCategories, fetchProducts } from '../../redux/actions/productAction';

import renderer from 'react-test-renderer';

jest.mock('../../redux/actions/productAction'); // Mock productAction

test('renders All button and category buttons', () => {
  render(
    <Provider store={store}>
      <CategoryBar />
    </Provider>
  );

  const allButton = screen.getByRole('button', { name: /All/i });
  const categoryButtons = screen.getAllByRole('button', { name: /category/i });

  expect(allButton).toBeInTheDocument();
  expect(categoryButtons).toHaveLength(2); // Adjust based on expected categories
});

test('dispatches actions on button clicks', () => {
  const dispatchMock = jest.fn();
  fetchCategories.mockReturnValue(dispatchMock);
  fetchProducts.mockReturnValue(dispatchMock);

  render(
    <Provider store={store}>
      <CategoryBar />
    </Provider>
  );

  const allButton = screen.getByRole('button', { name: /All/i });
  const categoryButton = screen.getByRole('button', { name: /category 1/i }); // Assuming a category named "category 1"

  userEvent.click(allButton);
  expect(fetchProducts).toHaveBeenCalledTimes(1);

  userEvent.click(categoryButton);
  expect(fetchProducts).toHaveBeenCalledTimes(2);
  expect(fetchProducts).toHaveBeenCalledWith('category 1');
});


test('renders correctly', () => {
  const tree = renderer.create(<CategoryBar categories={['category 1', 'category 2']} />).toJSON();
  expect(tree).toMatchSnapshot();
});
