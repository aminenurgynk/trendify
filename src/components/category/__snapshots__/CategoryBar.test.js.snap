// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryBar should render with given state from Redux store 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-16dwyet-MuiContainer-root"
  >
    <div
      aria-label="text button group"
      class="MuiButtonGroup-root MuiButtonGroup-text css-1pmhgw0-MuiButtonGroup-root"
      role="group"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedTextPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedTextPrimary MuiButtonGroup-firstButton css-1kqcr5r-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        All
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedTextPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedTextPrimary MuiButtonGroup-middleButton css-1kqcr5r-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        electronics
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedTextPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedTextPrimary MuiButtonGroup-middleButton css-1kqcr5r-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        books
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedTextPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedTextPrimary MuiButtonGroup-lastButton css-1kqcr5r-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        clothing
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
